File: LV1/1.5/ptice/ptice.cpp
#include <cstdio>

char a[300], b[300], g[300] = {0};

int max(int *score) {
    int m = -1;
    for (int i = 0; i < 3; i++) {
        if (score[i] > m) m = score[i];
    }
    return m;
}

void gen_ans() {
    for (int i = 0; i < 40; i++) {
        sprintf(a, "ABC%s", a);
        sprintf(b, "BABC%s", b);
        sprintf(g, "CCAABB%s", g);
    }
}

int main() {
    gen_ans();
    int n, score[3] = {0};
    char exam[110];
    char name[][10] = {"Adrian", "Bruno", "Goran"};

    scanf("%d", &n);
    scanf("%s", exam);

    for (int i = 0; i < n; i++) {
        if (a[i] == exam[i]) score[0]++;
        if (b[i] == exam[i]) score[1]++;
        if (g[i] == exam[i]) score[2]++;
    }
    int m = max(score);

    printf("%d\n", m);
    for (int i = 0; i < 3; i++) {
        if (score[i] == m) {
            printf("%s\n", name[i]);
        }
    }
}

File: LV1/1.5/ptice/ptice.hs
copyStr :: Int -> String -> String
copyStr n s = concat $ replicate n s

bNum :: Char -> Char -> Int
bNum a b = if a == b then 1 else 0

calcScore :: String -> [(Char, Char, Char)] -> [Int] -> [Int]
calcScore (e:es) ((a, b, c):ss) score =
  let res = [bNum a e, bNum b e, bNum c e]
  in  calcScore es ss $ zipWith (+) res score
calcScore _ _ score = score

main :: IO ()
main = do
  _    <- getLine
  exam <- getLine

  let scorePack = zip3 a b g
       where
        a = copyStr 40 "ABC"
        b = copyStr 25 "BABC"
        g = copyStr 20 "CCAABB"

  let name = ["Adrian", "Bruno", "Goran"]
      res  = calcScore exam scorePack [0, 0, 0]
      max' = maximum res

  print max'
  putStrLn . unlines . map snd $ filter (\(x, _) -> x == max') $ zip res name
